janusgraph-core/src/main/java/org/janusgraph/graphdb/database/cache/KCVSCacheInvalidationService.java

```diff
@@ -0,0 +1,68 @@
+// Copyright 2022 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.database.cache;
+
+import org.janusgraph.diskstorage.StaticBuffer;
+import org.janusgraph.diskstorage.keycolumnvalue.cache.KCVSCache;
+import org.janusgraph.graphdb.idmanagement.IDManager;
+
+import java.util.Collections;
+import java.util.List;
+
+public class KCVSCacheInvalidationService implements CacheInvalidationService{
+
+    private final List<KCVSCache> managedKCVSCaches;
+
+    private final IDManager idManager;
+
+    public KCVSCacheInvalidationService(List<KCVSCache> managedKCVSCaches, IDManager idManager) {
+        this.managedKCVSCaches = managedKCVSCaches;
+        this.idManager = idManager;
+    }
+
+    @Override
+    public void markVertexAsExpired(Long vertexId) {
+        StaticBuffer vertexIdKey = idManager.getKey(vertexId);
```

Comment:
Good point. I actually forgot to check that myself. I will double check and post an update

Reply:
You are right. This doesn't work for `indexStore`. Probably we can invalidate `edgeStore` only and `indexStore` should be invalidated by `key` somehow else.

Reply:
I guess I will change the interface of `CacheInvalidationService.java` to be able to expire either `edgeStore` or `indexStore` explicitly. For `edgeStore` we will be able to provide `vertexId` but for `indexStore` we will probably need to specify correct keys. Also, I guess a method like: `forceClearFullCache` could be helpful as well.

Reply:
@li-boxuan I'm think about the next interface for `CacheInvalidationService` below. What do you think about it?
```
/**
 * Marks specific vertex as expired.
 * It will make sure that any retrieved properties and edges associated with this vertex will be invalidated in vertex cache.
 * @param vertexId vertex id to expire
 */
void markVertexAsExpiredInEdgeStore(Long vertexId);

/**
 * Marks specific key as expired.
 * It will make sure that any retrieved properties and edges associated with this key will be invalidated in vertex cache.
 * @param key key to expire
 */
void markKeyAsExpiredInEdgeStore(StaticBuffer key);

/**
 * Marks specific key as expired.
 * It will make sure that any retrieved data associated with this key will be invalidated in index cache.
 * @param key key to expire
 */
void markKeyAsExpiredInIndexStore(StaticBuffer key);

/**
 * Instead of waiting for a probabilistic invalidation it triggers all cached queries scan and invalidation in vertex cache.
 * This will remove any cached expired data.
 */
void forceClearExpiredKeysInEdgeStoreCache();

/**
 * Instead of waiting for a probabilistic invalidation it triggers all cached queries scan and invalidation in index cache.
 * This will remove any cached expired data.
 */
void forceClearExpiredKeysInIndexStoreCache();

/**
 * Instead of waiting for a probabilistic invalidation it triggers all cached queries scan and invalidation for both edge store cache and index store cache.
 * This will remove any cached expired data.
 */
void forceClearExpiredCache();

/**
 * Marks a vertex as expired in edge store cache and triggers force clear of expired cache (i.e. {@link #forceClearExpiredKeysInEdgeStoreCache()})
 * @param vertexId vertex id to invalidate
 */
void forceInvalidateVertexInEdgeStoreCache(Long vertexId);

/**
 * Marks vertices as expired in edge store cache and triggers force clear of expired cache (i.e. {@link #forceClearExpiredKeysInEdgeStoreCache()})
 * @param vertexIds vertex ids to invalidate
 */
void forceInvalidateVerticesInEdgeStoreCache(Iterable<Long> vertexIds);

/**
 * Clears edge store cache fully
 */
void clearEdgeStoreCache();

/**
 * Clears index store cache fully
 */
void clearIndexStoreCache();

/**
 * Clears both edge store cache and index store cache fully.  
 * It is the same as calling {@link #clearEdgeStoreCache()} and {@link #clearIndexStoreCache()}
 */
void clearDBCache();
```

Reply:
Before commenting on that, I would like to understand more about the context. I am not sure how `CacheInvalidationService` will be used. Do you expect this to be used directly by users, or by JanusGraph internals?

Reply:
I would like to have both possibilities. I.e. I want to let users to expire / clear cache manually when they need it and also would like to add some invalidation strategy which can invalidate necessary cache records globally on mutation.
In this PR I want to have only first solution (I.e. for manual invalidation) and in the follow up PRs I would like to have a global invalidation strategy added

Reply:
A global invalidation strategy makes sense to me. I am hesitant about exposing this functionality to end users, mainly because it is non-trivial to evict the index cache. For example, when users do `graph.getDBCacheInvalidationService().forceInvalidateVertices(Arrays.asList(vertexId1, vertexId2)`, they would likely take it for granted that the index cache for `vertexId1` and `vertexId2` is also invalidated.

Taking a step back, I am wondering when this would be ever useful for end users to directly call these APIs. The internal usage (global invalidation) for sure is valuable but not sure if those APIs shall be exposed to end users.

Reply:
> Taking a step back, I am wondering when this would be ever useful for end users to directly call these APIs. The internal usage (global invalidation) for sure is valuable but not sure if those APIs shall be exposed to end users.

I agree that this invalidation service might not be obvious for users and maybe we need to hide it somewhere but I would still like to give advance users to access manual cache invalidation even so it's risky when they don't know how to work with cache invalidation. The reason for that is if users have their own log implementation with their own custom logic then can trigger cache invalidation on different nodes depending on their use case.
For example, if users already have custom log mechanism which logs only specific mutated relations (and not all relations), they can effectively invalidate them on different nodes. 
In our use case we don't want to invalidate all vertices when their properties are changed but we want to invalidate only those vertices which are served for clients. Any other vertex change is irrelevant to us and we don't want to log it or invalidate it.
If you feel that this service shouldn't be exposed to users publicly, I'm OK hiding it under `StandardJanusGraph`, but I would like to have a possibility to retrieve that service somehow without using Reflection API.
I guess if users really need to access DBCacheInvalidationService they can cast their JanusGraph to StandardJanusGraph are access it. I.e.:
```
StandardJanusGraph standardJanusGraph = (StandardJanusGraph) janusGraph;
DBCacheInvalidationService invalidationService = standardJanusGraph.getDBCacheInvalidationService();
``` 
What do you think about it?

Reply:
Got it, that sounds like a legitimate use case although very advanced. I am happy to have it exposed as public APIs, I was just hesitant about including it in the doc. If we write it in the doc, we need to be very careful and make sure people don't misuse it. Do you think JavaDoc itself would be sufficient?

Reply:
> Got it, that sounds like a legitimate use case although very advanced. I am happy to have it exposed as public APIs, I was just hesitant about including it in the doc. If we write it in the doc, we need to be very careful and make sure people don't misuse it. Do you think JavaDoc itself would be sufficient?

I think you are right. I will exclude it from the doc and I will try to improve JavaDoc with some examples maybe, so that users won't be confused about db-cache invalidation.